src/monitor/node_metadata.c

```diff
@@ -438,15 +438,20 @@ GetPrimaryOrDemotedNodeInGroup(char *formationId, int32 groupId)
 		return primaryNode;
 	}
 
-	/* maybe we have a primary that is draining or has been demoted? */
+	/*
+	 * Maybe we have a primary that is draining or has been demoted?
+	 * In case there are more than one of those, choose the one that is
+	 * currently being demoted.
+	 */
 	foreach(nodeCell, groupNodeList)
 	{
 		AutoFailoverNode *currentNode = (AutoFailoverNode *) lfirst(nodeCell);
 
-		if (StateBelongsToPrimary(currentNode->reportedState))
+		if (StateBelongsToPrimary(currentNode->reportedState) &&
+			(IsBeingDemotedPrimary(primaryNode) == false ||
+			IsDemotedPrimary(currentNode) == false))
 		{
 			primaryNode = currentNode;
-			break;
```

Comment:
Do we need to remove the break here? My understanding is that a short cut in this loop wouldn't have an impact, given the new tests, we find only one node and may return either the first or the last one then, it should be the same node, right?

Reply:
I have not tested it with the break to be honest. My understanding is the same as yours. I chose to keep the full loop under a 'belts and suspenders' argument, helped by the fact that the loops are generally very short anyways.